Sources/SentryCrash/Recording/SentryCrashBinaryImageCache.c

```diff
@@ -0,0 +1,115 @@
+#include "SentryCrashBinaryImageCache.h"
+#include "SentryCrashDynamicLinker.h"
+#include <mach-o/dyld.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define CACHE_SIZE_INCREMENT 100
+static SentryCrashBinaryImage *binaryImagesBuffer = NULL;
+static uint binaryImagesBufferLength;
+static int binaryImagesAmount;
+static pthread_mutex_t binaryImagesMutex = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+increaseBufferSize(void)
+{
+    pthread_mutex_lock(&binaryImagesMutex);
+    uint oldLength = binaryImagesBufferLength;
+    uint newLength = oldLength + CACHE_SIZE_INCREMENT;
+
+    SentryCrashBinaryImage *newBuffer = malloc(sizeof(SentryCrashBinaryImage) * newLength);
+    memcpy(newBuffer, binaryImagesBuffer, sizeof(SentryCrashBinaryImage) * oldLength);
+    free(binaryImagesBuffer);
+
+    binaryImagesBuffer = newBuffer;
+    binaryImagesBufferLength = newLength;
+    pthread_mutex_unlock(&binaryImagesMutex);
+}
+
+static void
+binaryImageAdded(const struct mach_header *header, intptr_t slide)
+{
+    if (binaryImagesBuffer == NULL) {
+        return;
+    }
+
+    Dl_info info;
+    if (!dladdr(header, &info) || info.dli_fname == NULL) {
+        return;
+    }
+
+    SentryCrashBinaryImage binaryImage = { 0 };
+    if (!sentrycrashdl_getBinaryImageForHeader(
+            (const void *)header, info.dli_fname, &binaryImage)) {
+        return;
+    }
+
+    if (binaryImagesAmount >= binaryImagesBufferLength) {
+        increaseBufferSize();
+    }
+
+    binaryImagesBuffer[binaryImagesAmount++] = binaryImage;
+}
+
+static void
+binaryImageRemoved(const struct mach_header *header, intptr_t slide)
+{
+    pthread_mutex_lock(&binaryImagesMutex);
+    int index = -1;
+    for (int i = 0; i < binaryImagesAmount; i++) {
+        if (binaryImagesBuffer[i].address == (uint64_t)header) {
+            index = i;
+            break;
+        }
+    }
+
+    if (index >= binaryImagesAmount) {
+        return;
+    }
+
+    if (index >= 0) {
+        int amountToMove = binaryImagesAmount - index + 1;
+        int sizeToMove = amountToMove * sizeof(SentryCrashBinaryImage);
+        void *startPosition = binaryImagesBuffer + ((index + 1) * sizeof(SentryCrashBinaryImage));
+        void *moveTo = startPosition - sizeof(SentryCrashBinaryImage);
+
+        memcmp(moveTo, startPosition, sizeToMove);
+        binaryImagesAmount--;
+    }
```

Comment:
Could we also just early return for all conditions and do the work below that? Also you'll need to release the lock before early returning.
```suggestion
    if (index < 0 || index >= binaryImagesAmount) {
    	pthread_mutex_unlock(&binaryImagesMutex);
        return;
    }

     int amountToMove = binaryImagesAmount - index + 1;
     int sizeToMove = amountToMove * sizeof(SentryCrashBinaryImage);
     void *startPosition = binaryImagesBuffer + ((index + 1) * sizeof(SentryCrashBinaryImage));
     void *moveTo = startPosition - sizeof(SentryCrashBinaryImage);

     memcmp(moveTo, startPosition, sizeToMove);
     binaryImagesAmount--;
```

Reply:
Side note, in the past I've used a C macro that works like Swift/Go `defer` that is helpful for things like releasing locks for all exit paths... see https://web.archive.org/web/20200307094401/https://oded.blog/2017/10/05/go-defer-in-cpp/. I've come close to adding it a few times for us, but it's C++ so brings in some complications.

Reply:
Is it possible to achieve this in C? The article shows a C++ suggestion.

Reply:
Probably, it would be interesting to try sometime!